src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DirectoryBasedStoreProvider.cs

```diff
@@ -0,0 +1,400 @@
+            // was requested.
+            OpenFlags writeFlags = openFlags & (OpenFlags.ReadWrite | OpenFlags.MaxAllowed);
+
+            if (writeFlags == OpenFlags.ReadOnly)
+            {
+                _readOnly = true;
+            }
+        }
+        
+        public void Dispose()
+        {
+            if (_watcher != null)
+            {
+                _watcher.Dispose();
+                _watcher = null;
+            }
+        }
+
+        public void FindAndCopyTo(
+            X509FindType findType,
+            object findValue,
+            bool validOnly,
+            X509Certificate2Collection collection)
+        {
+        }
+
+        public byte[] Export(X509ContentType contentType, string password)
+        {
+            // Export is for X509Certificate2Collections in their IStorePal guise,
+            // if someone wanted to export whole stores they'd need to do
+            // store.Certificates.Export(...), which would end up in the
+            // CollectionBackedStoreProvider.
+            Debug.Fail("Export was unexpected on a DirectoryBasedStore");
+            throw new InvalidOperationException();
+        }
+
+        public void CopyTo(X509Certificate2Collection collection)
+        {
+            Debug.Assert(collection != null);
+
+            // Copy the reference locally, any directory change operations
+            // will cause the field to be reset to null.
+            List<X509Certificate2> certificates = _certificates;
+
+            if (certificates == null)
+            {
+                // ReadDirectory will both load _certificates and return the answer, so this call
+                // will have stable results across multiple adds/deletes happening in parallel.
+                certificates = ReadDirectory();
+                Debug.Assert(certificates != null);
+            }
+
+            foreach (X509Certificate2 cert in certificates)
+            {
+                collection.Add(cert);
+            }
+        }
+
+        private List<X509Certificate2> ReadDirectory()
+        {
+            if (!Directory.Exists(_storePath))
+            {
+                // Don't assign the field here, because we don't have a FileSystemWatcher
+                // yet to tell us that something has been added.
+                return new List<X509Certificate2>(0);
+            }
+
+            List<X509Certificate2> certs = new List<X509Certificate2>();
+
+            foreach (string filePath in Directory.EnumerateFiles(_storePath, PfxWildcard))
+            {
+                X509Certificate2 cert;
+
+                try
+                {
+                    cert = new X509Certificate2(filePath);
+                }
+                catch (CryptographicException)
+                {
+                    // The file wasn't a certificate, move on to the next one.
+                    continue;
+                }
+
+                certs.Add(cert);
+            }
+
+            if (_watcher == null)
```

Comment:
Quite prossibly.  (prossibly: (n) The fine line between "possibly" and "probably")

I was hoping to avoid locking; but don't think I can.

Create FSW
Activate FSW
Read directory
FSW fires
assign _certs

Already have bad data.  So I guess I need to lock this method and the FSW event; and flip the activation to before the read.

Alternatively, I could dump the cache altogether; but multiple enumerations (such as multiple calls to `Find` (once I write that)) feels like it'll be more common than concurrent writes.